### ISSUE TITLE
2 strange code fragments

### ISSUE THREAD

**Issue by j-mattsson** [NVIDIA/Q2RTX#181]
(I don't have much experience with this codebase so maybe these are not bugs.)

1. https://github.com/NVIDIA/Q2RTX/blob/701cf318995c91486bb212114e8c8279287a5bb7/src/baseq2/g_phys.c#L582
`FRAMETIME` is the value `0.1f` of type `float`, `nextthink` is of type `int`. Wouldn't the addition result that is being assigned to `nextthink` be truncated to an `int` (ignoring the added `0.1f`) and therefore leave `nextthink` unchanged?

2. https://github.com/NVIDIA/Q2RTX/blob/701cf318995c91486bb212114e8c8279287a5bb7/src/refresh/models.c#L342-L349
`try_location` will be initialized in for-loop to either 0 (`TRY_MODEL_SRC_BASE`)  or 1 (`TRY_MODEL_SRC_GAME`). So if the if-statement (`try_location > 0`) on line 348 is true it means that `try_location` must be 1 and therefore the expression `try_location == TRY_MODEL_SRC_GAME` is always true and the ternary expression becomes pointless as it always returns `FS_PATH_GAME`.

**Comment by res2k** [comment]
1. With `nextthink` being measured in "frames" this should probably be +1. Looks like this is a bug in q2pro as well.
2. Yeah, it's redundant, but at least it's not harmful. Not entirely sure how it got in there, but I suspect it's due to copy&paste from some other location where `try_location` may have had the possibility to be negative.

**Comment by Rroff2** [comment]
Yeah it should be +1 - there may be other instances of stuff which hasn't converted over to the change in update rate functionality as it can be a bit complicated/non-standard how the original id code was implemented in that respect with some funny timer implementations.

Have to say I'm not entirely a fan of the change as I'm going to have to substantially rework my mod to support it hah (as I've also done some funny things with timer implementations out of time constraints) - should have some free time coming up to do it though.

**Comment by QwazyWabbitWOS** [comment]
Most of your statements about 1 are correct but I can find no cases where line 578, `if (part) {` is true. 
The backout never executes when I try to push anything, barrels, walls, monsters, boxes, etc. motion or barrels stuck against walls or boxes don't matter. One can safely replace `FRAMETIME` with 1, it will never matter for this function. This also explains why it pushing never changed behavior when `nextthink` was changed to type int.

Incidentally, it looks like when the `goto retry` was removed the code resetting nextthink became moot. The current form of that function dates back to 3.20 and when they took out the retry and the `obstacle->inuse` test the `} else {` clause at 594 became associated with the `if(part->blocked` clause instead of the `inuse` test but it's impossible to tell if that was added when they put in the `#if 0` line because I don't think we have a public version of the game code prior to 3.20. The function deserves a complete re-write but I doubt it would matter.

**Comment by QwazyWabbitWOS** [comment]
As for item 2, try_location is only "valid" when it is 1 or 0 and -1 is the termination of the loop.

**Comment by Rroff2** [comment]
As far back as the 3.14 source it says:

`	if (part)
	{
		// the move failed, bump all nextthink times and back out moves
		for (mv = ent ; mv ; mv=mv->teamchain)
		{
			if (mv->nextthink > 0)
				mv->nextthink += FRAMETIME;
		}

		// if the pusher has a "blocked" function, call it
		// otherwise, just stay in place until the obstacle is gone
		if (part->blocked)
			part->blocked (part, obstacle);
#if 0
		// if the pushed entity went away and the pusher is still there
		if (!obstacle->inuse && part->inuse)
			goto retry;
#endif
	}
	else
	{
		// the move succeeded, so call all think functions
		for (part = ent ; part ; part=part->teamchain)
		{
			SV_RunThink (part);
		}
	}
}`

If that is what you are referring to.

**Comment by QwazyWabbitWOS** [comment]
Yes, that's the section. The `retry` label is commented out and the `#if 0` takes away the reference to it. If this code goes back to 3.14 then that's good enough to tell us it's been that way a long time and no one has seen fit to deal with that function in any way. The `bug` introduced by making `nextthink` an int while not changing the increment from `FRAMETIME` to 1 won't be observable. I haven't found any scenarios yet where `if (part)` is not false so I am not certain that the lines that increment `nextthink` are ever executed.

**Comment by Rroff2** [comment]
Assuming that think is processed through SV_RunThink it would still function anyhow as long as the think time is above zero and less than a future time at next frame it will still get called - maybe more often than is desirable if there are any cases where it does work but not to an extent which would have any noticeable impact. Though I've not done a deep dive on what is going on.

**Comment by Paril** [comment]
> Most of your statements about 1 are correct but I can find no cases where line 578, `if (part) {` is true. The backout never executes when I try to push anything, barrels, walls, monsters, boxes, etc. motion or barrels stuck against walls or boxes don't matter. One can safely replace `FRAMETIME` with 1, it will never matter for this function. This also explains why pushing never changed behavior when `nextthink` was changed to type int.
> 
> Incidentally, it looks like when the `goto retry` was removed the code resetting nextthink became moot. The current form of that function dates back to 3.20 and when they took out the retry and the `obstacle->inuse` test the `} else {` clause at 594 became associated with the `if(part->blocked` clause instead of the `inuse` test but it's impossible to tell if that was added when they put in the `#if 0` line because I don't think we have a public version of the game code prior to 3.20. The function deserves a complete re-write but I doubt it would matter.

"part" is non-null when an object blocks a bmodel mover without being destroyed. Most objects get destroyed when they block something, but some don't or can't. It's pretty hard to run into this in vanilla but it does happen.

It is observable. Barrels aren't bmodels, so you won't see it from that.

**Comment by apanteleev** [comment]
The first issue (with the pusher) has been fixed in Q2PRO and merged here: https://github.com/NVIDIA/Q2RTX/commit/95379a6a31ffd1d4ae226222c26cb582a85d491b

**Comment by apanteleev** [comment]
The project is discontinued; no new features or fixes will be implemented here. Closing.
